Updated the lock statement article

pkulikov opened the pull request:
Updated the [lock statement](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/lock-statement) article. This update is heavily based on the [specification](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#the-lock-statement)

A new code example is dotnet/samples#262

Contributes to #4992 and #3819

@mikedn as you mentioned that
>The related Lock statement too has flaws.

your review and feedback will be appreciated.

pkulikov committed:
Updated the lock statement article

rpetrusha commented:
Closing and reopening to begin new build.

rpetrusha commented:
Build still can't find code example. Closing and reopening...

rpetrusha reviewed (changes requested):
This is a really good revision, @pkulikov. I've left a couple of comments. Also, I'll try closing and reopening one more time -- for some reason, my two previous closings and reopenings haven't begun a new build.

rpetrusha commented:
Closing and reopening to begin new build.

mikedn commented on docs/csharp/language-reference/keywords/lock-statement.md:
````diff
@@ -11,76 +11,69 @@ ms.assetid: 656da1a4-707e-4ef6-9c6e-6d13b646af42
+```
 
-The `lock` keyword calls <xref:System.Threading.Monitor.Enter%2A> at the start of the block and <xref:System.Threading.Monitor.Exit%2A> at the end of the block. A <xref:System.Threading.ThreadInterruptedException> is thrown if <xref:System.Threading.Thread.Interrupt%2A> interrupts a thread that is waiting to enter a `lock` statement.
+except that `x` is only evaluated once.
````
Might be better to just show that in the example - add `object __lockObj = x;` and use `__lockObj` in the monitor calls.

mikedn commented on docs/csharp/language-reference/keywords/lock-statement.md:
```diff
@@ -11,76 +11,69 @@ ms.assetid: 656da1a4-707e-4ef6-9c6e-6d13b646af42
+## Remarks
 
-Best practice is to define a `private` object to lock on, or a `private static` object variable to protect data common to all instances.
+Avoid locking on publicly accessible instances, as it might result in lock contention or deadlock. Consider the following constructs:
```
Using "publicly accessible" in relation to object instances is potentially confusing, especially that one of the cases below also uses that in relation to types. Though I'm not really sure what terminology works best here. Maybe it should say something like:

Avoid using the same lock object instance for different shared resources, as it might result in deadlock or lock contention. Use a dedicated object instance (`private readonly object listLock = new object()`) or another instance that is unlikely to be used as a lock object by unrelated parts of the code. This in general precludes the use of `this` (potentially used as a lock by the callers), `Type` instances (be it obtained through `typeof` or reflection) or string instances, including string literals.

Might be a bit convoluted but anyway, something along those lines.

pkulikov replied:
@mikedn thanks, this comment helped a lot.

mikedn commented on docs/csharp/language-reference/keywords/lock-statement.md:
```diff
@@ -11,76 +11,69 @@ ms.assetid: 656da1a4-707e-4ef6-9c6e-6d13b646af42
+- `lock(this)` is a problem if the instance referenced by `this` can be accessed publicly.
 
-## Example - Threads without locking
+- `lock(typeof(MyType))` is a problem if `MyType` is publicly accessible.
```
> is a problem if `MyType` is publicly accessible

Probably it's best not to get into details. Any `Type` instance is "accesible", even if not public. There's reflection for that.

mikedn commented on docs/csharp/language-reference/keywords/lock-statement.md:
```diff
@@ -11,76 +11,69 @@ ms.assetid: 656da1a4-707e-4ef6-9c6e-6d13b646af42
+- `lock(typeof(MyType))` is a problem if `MyType` is publicly accessible.
 
-The following sample shows a simple use of threads without locking in C#:
+- `lock("myLock")` is a problem because any other code in the process using the same string, would share the same lock.
```
This applies to all string instances to do posibility of string interning. It is also possible go in the other direction and end up not sharing a lock when you want it - AFAIR it's possible that the same string literal to yield different string instances in different assemblies.

mikedn commented:
The Remarks section could use some rewording but overall this is much better than before. Thanks!

pkulikov committed:
Addressed feedback

pkulikov committed:
Addressed feedback

pkulikov committed:
One more read

pkulikov commented:
@rpetrusha @mikedn thank you very much for your review and feedback. It improved the text a lot. I've addressed your feedback and additionally merged one paragraph into the first one. The **Remarks** section has become better after @mikedn's suggestion. Please check the updated version.

rpetrusha commented:
Thank you so much, @pkulikov. This article is enormously improved as a result of your efforts. I'll merge now.
